FStarLang/FStar
# Build F* with `dune` instead of `ocamlbuild`

# tl;dr: What changes for the user

* `fstarlib`, `fstar-tactics-lib` and `fstar-compiler-lib` all become `fstar.lib`
* `OCAMLPATH=$FSTAR_HOME/bin` becomes `OCAMLPATH=$FSTAR_HOME/lib`
  * `$FSTAR_HOME/bin/fstarlib.cmx*` becomes `$FSTAR_HOME/lib/fstar/lib/fstar_lib.cmx*`
* the tactic library is now statically linked with `fstar.exe`
* in the binary package, the standard library is now in `lib/fstar` instead of `ulib` (this was already the case with the OPAM package or `make install`)
* compiling F* programs extracted to OCaml with the F* binary package is no longer supported, please install F* via opam instead (with `opam pin fstar --dev-repo`)

# The problem

Currently, the OCaml port of F*, both for Windows and Linux, relies on ocamlbuild, which leads to the following issues:

* no parallelism
* poor incrementality
* ocamlbuild is explicitly dubbed as a "legacy" build system for OCaml, undergoing deprecation in favor of Dune (cf. https://github.com/ocaml/ocamlbuild)

# The solution: this PR

This PR builds F* with `dune` instead of `ocamlbuild`. To this end:

* a whole snapshot of F*, including `ulib`, is now included, currently in the `ocaml` subdirectory. Now, to build F*, it is only necessary to build that snapshot and verify `ulib`. This is what the default `make` rule of the root Makefile now does. No extraction is needed as long as nothing in the F* sources or in `ulib` is changed, which is the case for most users. (In particular, lax-typechecking of `ulib` is no longer needed!) As of now, the snapshot includes `FStar_Parser_Parse.ml` file as generated by menhir/ocamlyacc, but no version of `parse.mly` file (so that we don't need to instruct dune to call menhir or ocamlyacc)
* `fstar-lib`, `fstar-tactics-lib` and `fstar-compiler-lib` are now all merged into a single package, `fstar.lib`, whose OCaml sources (including the generated and extracted files) are in `ocaml/fstar-lib`, which is now statically linked with the F* compiler. No need to dynamically load any "F* tactic library" anymore! No need to compute a list of "leftovers between fstarlib and fstartaclib" anymore! The library is now installed into `lib/fstar`, instead of a subdirectory of `bin`.
  (With this solution, every application built with F* needs to be linked with the whole F* compiler/typechecker, etc., which may cause bloating of the produced binary executable. As discussed with @nikswamy  and @mtzguido , we may need to change the way `--codegen Plugin` works, by making sure it produces a library disjoint from `--codegen OCaml`, at least at the level of the module name.)
* all handwritten OCaml files are now in the `ocaml` subdirectory, instead of `src/*/ml`, `ulib/ml` or `ulib/tactics_ml`, with the exception of `parse.mly`. The entrypoint of `fstar.exe` is in `ocaml/fstar`, that of `fstar_tests.exe` is in `ocaml/fstar-tests`; everything else is in `ocaml/fstar-lib`. However, we distinguish them from the generated files, which are in `ocaml/fstar-lib/generated` (instead of `src/ocaml-output`). This is both to make sure there is only one copy of each file, and to satisfy `dune`'s requirement that all OCaml files shall be in the same directory hierarchy.

## How to test: 3 `make` rules

* To build F*, just do `make` 
* With an existing F*, to update the snapshot in case of any change in `src` or `ulib`, just do `make dune-bootstrap`
  This rule now supports incrementality.
* If needed, the rule `make dune-full-bootstrap` removes all generated files from the snapshot before generating them again. This shouldn't be necessary, except maybe if a file is renamed and we want to make sure the old file is removed.

All those 3 rules support parallelism.

## Remaining tasks before merging this PR

* Fix the remaining state-handling issues (behavior discrepancies between the executables compiled by dune and ocamlbuild) (see bc0072444c45747a3bc2564b6e5ebd37057f9e24):
  * [x] An "error 147" message disappeared from `tests/ide/emacs/search.cons-snoc`
  * [x] uvar numbers shifted in Bug2010, Bug2021, OccursCheckOnArrows.
  * As part of this PR, I fixed a bug in the pretty-printer, where some internal state (`unfold_tuples`) was not properly restored, cf. 0b02e4850ce81b7c2c1dea1493e349ee28926e7f . This bug was causing a diff in `tests/prettyprinting/Tuples`, with extra parentheses in the dune build as opposed to the ocambuild build.
* [x] Remove now-unused, useless attempts to dynamically load the old fstartaclib.cmxs (easy)
* [x] Re-enable the snapshot diff check in the CI. To do that, we need to:
  * [x] Fix the elaboration of static quotations, which are currently printing the full name of the file where the file is built, thus introducing a discrepancy between the committer's machine and the CI container
* [x] Fix the `make install` rule (for binary packages and opam)
* [x] Clarify the opam package: probably `ocaml/fstar.opam` is not needed, but we can benefit from the new snapshot to reduce the dependencies in the existing `fstar.opam`
* [x] For `make package`, rebuild and include the F# snapshot of ulib
* [ ] Fix the release infra
* [x] Fix the `nix` rules
* [x] Check that everest still works (this should be fine)

# What this pull request DOES NOT do

* Instruct `dune` to generate the F* parser using menhir and/or ocamlyacc
* Build each ocaml extraction example with `dune` instead of `ocamlbuild` (`hello/multifile` already does, but not `sample_project`)
* Extract the F* sources without `--lax`, benefitting from the `.checked` files of `ulib`
* Move the standard library from `ulib` to `lib/fstar` in the F* sources
* Verify `ulib` with `dune` instead of `make` (this would replace `make` with `dune` as the build driver for F*)
